OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs

```diff
@@ -0,0 +1,268 @@
+
+		public object Create(ActorInitializer init) { return new EditorActorLayer(init.Self, this); }
+	}
+
+	public class EditorActorLayer : IWorldLoaded, ITickRender, IRender, IRadarSignature, ICreatePlayers
+	{
+		readonly EditorActorLayerInfo info;
+		readonly Dictionary<string, EditorActorPreview> previews = new Dictionary<string, EditorActorPreview>();
+		readonly Dictionary<CPos, List<EditorActorPreview>> cellMap = new Dictionary<CPos, List<EditorActorPreview>>();
+
+		SpatiallyPartitioned<EditorActorPreview> screenMap;
+		WorldRenderer worldRenderer;
+
+		public MapPlayers Players { get; private set; }
+
+		public EditorActorLayer(Actor self, EditorActorLayerInfo info)
+		{
+			this.info = info;
+		}
+
+		public void CreatePlayers(World w)
+		{
+			if (w.Type != WorldType.Editor)
+				return;
+
+			Players = new MapPlayers(w.Map.PlayerDefinitions);
+		}
+
+		public void WorldLoaded(World world, WorldRenderer wr)
+		{
+			if (world.Type != WorldType.Editor)
+				return;
+
+			worldRenderer = wr;
+
+			foreach (var pr in Players.Players.Values)
+				wr.UpdatePalettesForPlayer(pr.Name, pr.Color, false);
+
+			var ts = Game.ModData.Manifest.TileSize;
+			var width = world.Map.MapSize.X * ts.Width;
+			var height = world.Map.MapSize.Y * ts.Height;
+			screenMap = new SpatiallyPartitioned<EditorActorPreview>(width, height, info.BinSize);
+
+			foreach (var kv in world.Map.ActorDefinitions)
+				Add(kv.Key, new ActorReference(kv.Value.Value, kv.Value.ToDictionary()), true);
+
+			// Update neighbours in one pass
+			foreach (var p in previews.Values)
+				UpdateNeighbours(p.Footprint);
+		}
+
+		public void TickRender(WorldRenderer wr, Actor self)
+		{
+			if (wr.World.Type != WorldType.Editor)
+				return;
+
+			foreach (var kv in previews.Values)
+				kv.Tick();
+		}
+
+		static readonly IEnumerable<IRenderable> NoRenderables = Enumerable.Empty<IRenderable>();
+		public virtual IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr)
+		{
+			if (wr.World.Type != WorldType.Editor)
+				return NoRenderables;
+
+			return PreviewsInBox(wr.Viewport.TopLeft, wr.Viewport.BottomRight)
+				.SelectMany(p => p.Render());
+		}
+
+		public EditorActorPreview Add(ActorReference reference) { return Add(NextActorName(), reference); }
+
+		EditorActorPreview Add(string id, ActorReference reference, bool initialSetup = false)
+		{
+			var owner = Players.Players[reference.InitDict.Get<OwnerInit>().PlayerName];
+
+			var preview = new EditorActorPreview(worldRenderer, id, reference, owner);
+			previews.Add(id, preview);
+			screenMap.Add(preview, preview.Bounds);
+
+			foreach (var kv in preview.Footprint)
+			{
+				List<EditorActorPreview> list;
+				if (!cellMap.TryGetValue(kv.Key, out list))
+				{
+					list = new List<EditorActorPreview>();
+					cellMap.Add(kv.Key, list);
+				}
+
+				list.Add(preview);
+			}
+
+			if (!initialSetup)
+			{
+				UpdateNeighbours(preview.Footprint);
+
+				if (reference.Type == "mpspawn")
```

Comment:
The multiple references to this hardcoded string means it ought to be exposed to yaml, IMO.